README.md
![logo](https://github.com/senhasegura/ansible-pam-collection/blob/main/docs/imgs/senhasegura-logo.jpg?raw=true)


## senhasegura Privileged Access Management Ansible Collection

senhasegura PAM Ansible Collection provides ansible modules for interacting with senhasegura's PAM module

Hosted in [galaxy.ansible.com/senhasegura](https://galaxy.ansible.com/senhasegura)

*************

### Requirements

  - senhasegura with PAM module enabled

---

### Installation

In your terminal, run the following command

	ansible-galaxy collection install senhasegura.pam

### Modules

**senhasegura_device**

With this module you can create, update and deactivate devices in senhasegura's PAM module

[Detailed documentation](https://github.com/senhasegura/ansible-pam-collection/blob/main/docs/senhasegura_device.md)

**senhasegura_credential**

With this module you can create, update and deactivate credentials in senhasegura's PAM module

[Detailed documentation](https://github.com/senhasegura/ansible-pam-collection/blob/main/docs/senhasegura_credential.md)

**senhasegura_credential_info**

With this module you can get informations about a credential, including your password

[Detailed documentation](https://github.com/senhasegura/ansible-pam-collection/blob/main/docs/senhasegura_credential_info.md)

---

## Author information

- Lucas Fraga (@lfraga)
- James Miranda (@jameswpm)



CHANGELOG.md
# Changelog

## [1.0.0] - 2021-02-15

### Added

- The following modules have been added to [senhasegura-pam-collection](https://galaxy.ansible.com/senhasegura/pam)
	- senhasegura_device
	- senhasegura_credential
	- senhasegura_credential_info

[1.0.0]: https://github.com/senhasegura/ansible-pam-collection/tree/v1.0.0



plugins/module_utils/iso.py
#!/usr/bin/python
# Copyright: (c) 2017, Ansible Project
# GNU General Public License v3.0+
# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
from ansible.module_utils._text import to_text
from ansible.module_utils.six.moves import http_client as httplib
from ansible.module_utils.six.moves.urllib.error import HTTPError

import requests
import json

__metaclass__ = type

def authenticate(module):
	"""
		Perform the Oauth2 authentication at senhasegura
		server to get the token for execute main actions
	"""

	url = '{}/iso/oauth2/token'.format(module.params["system_url"])

	data = {
		'grant_type': 'client_credentials',
		'client_id': module.params["client_id"],
		'client_secret': module.params["client_secret"]
	}


	try:

		resp = requests.post(url, data=data, verify=module.params["validate_certs"])

		if resp.json()['access_token'] is None:
			module.fail_json(
				msg="Could not authenticate with the server, check your credentials"
			)

		return resp.json()['access_token']

	except (HTTPError, httplib.HTTPException) as http_exception:

		module.fail_json(
			msg=("Could not authenticate with the server, check your credentials."
				 "Please validate parameters provided."
				 "\n*** end_point=%s\n ==> %s" % (url, to_text(http_exception))),
			payload=data,
			status_code=http_exception.code)

	except Exception as unknown_exception:

		module.fail_json(
			msg=("Error while performing OAuth2 authentication."
				 "\n*** end_point=%s\n%s" % (url, to_text(unknown_exception))),
			payload=data,
			status_code=-1)




def iso_request(module, url, method="GET", data={}, headers={}, required_http_code=[200]):

	try:
		if method == "POST":
			resp = requests.post(url, headers=headers, data=data, json=json.dumps(data), verify=module.params["validate_certs"])
		elif method == "DELETE":
			resp = requests.delete(url, headers=headers, data=data, verify=module.params["validate_certs"])
		elif method == "GET":
			resp = requests.get(url, headers=headers, verify=module.params["validate_certs"])

		if resp.status_code not in required_http_code:
			module.fail_json(msg="Error: HTTP {}".format(resp.status_code), payload=data, status_code=-1)

		return resp

	except (HTTPError, httplib.HTTPException) as http_exception:

		module.fail_json(
			msg=("Error sending API request"
				 "Please validate parameters provided."
				 "\n*** end_point=%s\n ==> %s" % (url, to_text(http_exception))),
			payload=data,
			status_code=http_exception.code)

	except Exception as unknown_exception:

		module.fail_json(
			msg=("Unknown error in API request"
				 "\n*** end_point=%s\n%s" % (url, to_text(unknown_exception))),
			payload=data,
			STatus_code=-1)


plugins/modules/senhasegura_device.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2021, Ansible Project
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

from ansible.module_utils.basic import AnsibleModule, env_fallback

from ansible_collections.senhasegura.pam.plugins.module_utils.iso import authenticate, iso_request

__metaclass__ = type

DOCUMENTATION = r'''
---
module: senhasegura_device
short_description: Module for management devices in senhasegura PAM
author:
  - Lucas Fraga (@lfraga)
description:
    - Authenticates in senhasegura PAM to manage devices
notes:
    - Check mode not supported

options:
    system_url:
        type: str
        description:
            - senhasegura's environment URL used for authentication, SENHASEGURA_URL environment variable may also be used
    client_id:
        type: str
        description:
            - A2A OAuth2 client_id, SENHASEGURA_CLIENT_ID environment variable may also be used
    client_secret:
        type: str
        description:
            - A2A OAuth2 client_secret, SENHASEGURA_CLIENT_SECRET environment variable may also be used
    state:
        description:
            - Whether the device should exist or not,
        type: str
        choices: [ absent, present ]
        default: present
    hostname:
        required: true
        type: str
        description:
            - Hostname of device
    address:
        required: true
        type: str
        description: IP or DNS
    type:
        description:
            - Type of device, if non-existent, will be created.
            - Required when state is present
        type: str
    vendor:
        description:
            - Vendor of device, if non-existent, will be created.
            - Required when state is present
        type: str
    model:
        description:
            - Model of device, if non-existent, will be created.
            - Required when state is present
        type: str
    site:
        description:
            - Site of device.
            - Required when state is present
        type: str
    domain:
        description:
            - The device domain.
        type: str
    connectivities:
        description:
            - The connectivity protocols on device in format <Protocol>:<Port>
            - Example 'SSH:22'
        type: list
    tags:
        description:
            - The device tags.
        type: list
    validate_certs:
        description:
            - Whether to validate or not the HTTPS certificate
        type: boolean
'''

EXAMPLES = r'''
- name: Absent example-001 device
  senhasegura.pam.senhasegura_device:
    state: absent
    hostname: example-001
    address: 192.168.10.10


- name: Ensure example-002 device
  senhasegura.pam.senhasegura_device:
    state: present
    hostname: example-002
    address: 192.168.10.20
    type: Server
    vendor: Red Hat
    model: Red Hat Enterprise Linux
    site: Cloud
    connectivities: ['SSH:22', 'HTTPS:443']
    tags: ['rhel', 'cloud']

- name: Ensure windows-001 device
  senhasegura.pam.senhasegura_device:
    state: present
    hostname: windows-001
    address: 192.168.10.30
    type: Server
    vendor: Microsoft
    model: Windows Server
    site: On-premises
    connectivities: ['RDP:3389']
    tags: ['windows', 'domain']
    domain: CORP

'''

RETURN = r'''
changed:
    description: Whether there was a change done.
    type: bool
    returned: always
senhasegura_device:
    description: Dictionary containing result properties.
    returned: always
    type: dict
    sample:
        result:
            description: device properties, based on senhasegura informations
            type: complex
            returned: success
status_code:
    description: Result HTTP Status code
    returned: always
    type: int
    sample: 200
'''


def ensure_device(module, authentication_token):
    params = module.params

    headers = {"Authorization": 'Bearer {}'.format(authentication_token)}

    url = params['system_url'] + '/iso/pam/device'

    payload = {
        "ip": params["address"],
        "hostname": params["hostname"],
        "type": params["type"],
        "vendor": params["vendor"],
        "model": params["model"],
        "site": params["site"]
    }

    if params['domain'] is not None:
        payload['device_domain'] = params['domain']

    if params['tags'] is not None:
        payload['device_tags'] = ','.join(params['tags'])

    if params["connectivities"] is not None:
        payload['connectivities'] = ','.join(params["connectivities"])

    r = iso_request(module, url, method="POST", headers=headers,
                    data=payload, required_http_code=[200, 201])

    result = {"result": r.json()}

    # Check idempotency
    if "changed" in result["result"]["response"]:
        if result["result"]["response"]["changed"] == "true":
            changed = True
        else:
            changed = False
    else:
        changed = True

    return (changed, result, r.status_code)


def ensure_absent_device(module, authentication_token):
    params = module.params

    url = params['system_url'] + '/iso/pam/device'
    headers = {"Authorization": 'Bearer {}'.format(authentication_token)}

    payload = {}

    url += '/{}'.format(params['hostname'])
    payload['device'] = params['hostname']

    r = iso_request(module, url, method="DELETE", headers=headers,
                    data=payload, required_http_code=[200, 400, 404])

    if r.status_code == 400:
        # Device deactivated or not found
        changed = False
    elif r.status_code == 200:
        # Device deactivated
        changed = True

    result = {"result": r.json()}
    return (changed, result, r.status_code)


def main():

    module_args = dict(
        system_url=dict(type='str', required=True, fallback=(env_fallback, ['SENHASEGURA_URL'])),
        client_id=dict(type='str', no_log=True, required=True, fallback=(env_fallback, ['SENHASEGURA_CLIENT_ID'])),
        client_secret=dict(type='str', no_log=True, required=True, fallback=(env_fallback, ['SENHASEGURA_CLIENT_SECRET'])),
        state=dict(type='str', default='present',
                   choices=['absent', 'present']),
        hostname=dict(type='str'),
        address=dict(type='str'),
        type=dict(type='str'),
        vendor=dict(type='str'),
        model=dict(type='str'),
        site=dict(type='str'),
        domain=dict(type='str'),
        tags=dict(type='list'),
        connectivities=dict(type='list'),
        validate_certs=dict(type='bool', default=True)
    )

    required_if = [
        ["state", "present", ["address", "hostname", "site", "model", "vendor", "type"]],
        ["state", "absent", ["address", "hostname"]]
    ]

    module = AnsibleModule(
        argument_spec=module_args,
        required_if=required_if,
        supports_check_mode=False
    )

    authentication_token = authenticate(module)

    if (module.params['state'] == "present"):
        (changed, result, status_code) = ensure_device(
            module, authentication_token)

    elif (module.params['state'] == "absent"):
        (changed, result, status_code) = ensure_absent_device(
            module, authentication_token)

    module.exit_json(
        changed=changed,
        senhasegura_device=result,
        status_code=status_code)


if __name__ == '__main__':
    main()


plugins/modules/senhasegura_credential_info.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2021, Ansible Project
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

from ansible.module_utils.basic import AnsibleModule, env_fallback

from ansible_collections.senhasegura.pam.plugins.module_utils.iso import authenticate, iso_request

__metaclass__ = type

DOCUMENTATION = r'''
---
module: senhasegura_credential_info
short_description: Module for get information about credentials in senhasegura PAM
author:
  - Lucas Fraga (@lfraga)
description:
    - Authenticates in senhasegura PAM to get credentials
notes:
    - Check mode not supported

options:
    system_url:
        type: str
        description:
            - senhasegura's environment URL used for authentication, SENHASEGURA_URL environment variable may also be used
    client_id:
        type: str
        description:
            - A2A OAuth2 client_id, SENHASEGURA_CLIENT_ID environment variable may also be used
    client_secret:
        type: str
        description:
            - A2A OAuth2 client_secret, SENHASEGURA_CLIENT_SECRET environment variable may also be used
    identifier:
        type: str
        description:
            - The credential identifier to get information
    id_credential:
        type: str
        description:
            - The credential ID to get information
    validate_certs:
        description:
            - Whether to validate or not the HTTPS certificate
        type: bool
'''

EXAMPLES = r'''
- name: Get informations about credential with identifier 'example_credential'
  senhasegura.pam.senhasegura_credential_info:
    identifier: example_credential
  register: senhasegura_example_credential

- name: Get informations about credential with id_credential '9172'
  senhasegura.pam.senhasegura_credential_info:
    id_credential: 9172
  register: senhasegura_9172_credential

'''

RETURN = r'''
changed:
    description: Whether there was a change done.
    type: bool
    returned: always
senhasegura:
    description: Dictionary containing result properties.
    returned: always
    type: dict
    sample:
        credential:
            description: credentials properties, based on senhasegura informations
            type: complex
            returned: success
status_code:
    description: Result HTTP Status code
    returned: always
    type: int
    sample: 200
'''


def credential_info(module, authentication_token):

    params = module.params

    url = params["system_url"] + '/iso/pam/credential/'

    if params['id_credential'] is not None:
        url += '{}?credential={}'.format(params['id_credential'], params['id_credential'])
    elif params['identifier'] is not None:
        url += '{}?credential={}'.format(params['identifier'], params['identifier'])

    headers = {'Content-Type': 'application/json',
               "Authorization": 'Bearer {}'.format(authentication_token)}


    r = iso_request(module, url, method="GET", headers=headers, required_http_code=[200])

    result = r.json()
    return (True, result, r.status_code)


def main():
    module_args = dict(
        system_url        = dict(type='str', required=True, fallback=(env_fallback, ['SENHASEGURA_URL'])),
        client_id         = dict(type='str', no_log=True, required=True, fallback=(env_fallback, ['SENHASEGURA_CLIENT_ID'])),
        client_secret     = dict(type='str', no_log=True, required=True, fallback=(env_fallback, ['SENHASEGURA_CLIENT_SECRET'])),
        id_credential     = dict(type='int'),
        identifier        = dict(type='str'),
        validate_certs    = dict(type='bool', default=True)
    )


    mutually_exclusive = [
        ["id_credential", "identifier"]
    ]

    module = AnsibleModule(
        argument_spec       = module_args,
        mutually_exclusive  = mutually_exclusive,
        supports_check_mode = False
    )

    authentication_token = authenticate(module)

    (changed, result, status_code) = credential_info(module, authentication_token)

    module.exit_json(
        changed=changed,
        senhasegura=result,
        status_code=status_code)


if __name__ == '__main__':
    main()


plugins/modules/senhasegura_credential.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2021, Ansible Project
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

from ansible.module_utils.basic import AnsibleModule, env_fallback

from ansible_collections.senhasegura.pam.plugins.module_utils.iso import authenticate, iso_request

__metaclass__ = type

DOCUMENTATION = r'''
---
module: senhasegura_credential
short_description: Module for management credentials in senhasegura PAM
author:
  - Lucas Fraga (@lfraga)
  - James Miranda (@jameswpm)
description:
    - Authenticates in senhasegura PAM to manage credentials
notes:
    - Check mode not supported

options:
    system_url:
        type: str
        description:
            - senhasegura's environment URL used for authentication, SENHASEGURA_URL environment variable may also be used
    client_id:
        type: str
        description:
            - A2A OAuth2 client_id, SENHASEGURA_CLIENT_ID environment variable may also be used
    client_secret:
        type: str
        description:
            - A2A OAuth2 client_secret, SENHASEGURA_CLIENT_SECRET environment variable may also be used
    state:
        description:
            - Whether the credential should exist or not,
        type: str
        choices: [ absent, present ]
        default: present
    identifier:
        required: true
        type: str
        description:
            - credential identifier for management
    device_hostname:
        type: str
        description:
            - Device hostname to link credential
            - Required when state is present
    device_address:
        type: str
        description:
            - Device address to link credential
            - Required when state is present
    type:
        description:
            - Type of credential
            - Required when state is present
        type: str
        choices: [ Local User, Local administrator, Domain user ]
        default: Local User
    username:
        type: str
        description:
            - Credential username
            - Required when state is present
    password:
        description:
            - Credential password
            - This password must meet the password policy requirements
        type: str
    domain:
        description:
            - Domain name, It needs to be previously registered in senhasegura
        type: str
    tags:
        description:
            - Credential tags
        type: list
    additional_info:
        description:
            - Credential additional information
        type: str
    parent_credential:
        description:
            - Parent credential numeric identifier
        type: int
    validate_certs:
        description:
            - Whether to validate or not the HTTPS certificate
        type: boolean
'''

EXAMPLES = r'''
- name: Ensure credential 'ansible' in device 'example-001'
  senhasegura.pam.senhasegura_credential:
    state: present
    identifier: example_credential
    device_hostname: example-001
    device_address: 192.168.10.10
    type: Local user
    username: ansible
    password: your_secret_password
    tags:
      - ansible
      - example
    additional_info: 'Created by Ansible'

- name: Ensure domain credential
  senhasegura.pam.senhasegura_credential:
    state: present
    identifier: domain_credential
    device_hostname: example-002
    device_address: 192.168.10.20
    type: Domain user
    username: ansible
    password: your_secret_password
    domain: CORP
    tags:
      - domain

- name: Inactivate credential with identifier 'domain_credential'
  senhasegura.pam.senhasegura_credential:
    state: absent
    identifier: domain_credential

'''

RETURN = r'''
changed:
    description: Whether there was a change done.
    type: bool
    returned: always
senhasegura_credential:
    description: Dictionary containing result properties.
    returned: always
    type: dict
    sample:
        result:
            description: credentials properties, based on senhasegura informations
            type: complex
            returned: success
status_code:
    description: Result HTTP Status code
    returned: always
    type: int
    sample: 200
'''


def ensure_credential(module, authentication_token):
    """
        Make the request to save the credential, it is possible that will be created
        or edited if a valid identifier or pair username with ip is sent
    """

    params = module.params

    # Required parameters
    payload = {
        "ip": params["device_address"],
        "hostname": params["device_hostname"],
        "type": params["type"],
        "username": params["username"],
        "identifier": params["identifier"]
    }


    url = params['system_url'] + '/iso/pam/credential'

    headers = {"Authorization": 'Bearer {}'.format(authentication_token)}

    # Optional parameters
    if params['password']:
        payload['content'] = params['password']

    if params['domain']:
        payload['domain'] = params['domain']

    if params['tags'] is not None:
        payload['tags'] = ','.join(params['tags'])

    if params['additional_info']:
        payload['additional'] = params['additional_info']

    if params['parent_credential']:
        payload['parent_password'] = params['parent_credential']


    r = iso_request(module, url, method="POST", headers=headers, data=payload, required_http_code=[200, 201])

    # Check idempotency
    result = {"result": r.json()}

    if "changed" in result["result"]["response"]:
        if result["result"]["response"]["changed"] == "true":
            changed = True
        else:
            changed = False
    else:
        changed = True

    return (changed, result, r.status_code)


def ensure_absent_credential(module, authentication_token):
    """
        Performs the request to delete the credential
        In deletion, the credential to remove can be fetched by id, identifier or
        pair username with ip
    """
    params = module.params

    url = params['system_url'] + '/iso/pam/credential/{}'.format(params['identifier'])

    payload = {
        "identifier": params["identifier"]
    }

    headers = {'Content-Type': 'application/json',
               "Authorization": 'Bearer {}'.format(authentication_token)}


    r = iso_request(module, url, method="DELETE", headers=headers, data=payload, required_http_code=[200,201,400])

    # Check idempotency status
    if r.status_code == 400:
        changed = False
    else:
        changed = True

    result = {"result": r.json()}
    return (changed, result, r.status_code)


def main():
    module_args = dict(
        system_url        = dict(type='str', required=True, fallback=(env_fallback, ['SENHASEGURA_URL'])),
        client_id         = dict(type='str', no_log=True, required=True, fallback=(env_fallback, ['SENHASEGURA_CLIENT_ID'])),
        client_secret     = dict(type='str', no_log=True, required=True, fallback=(env_fallback, ['SENHASEGURA_CLIENT_SECRET'])),
        state             = dict(type='str', default='present', choices=['absent', 'present']),
        identifier        = dict(type='str'),
        device_hostname   = dict(type='str'),
        device_address    = dict(type='str'),
        type              = dict(type='str', default='Local User', choices=['Local User', 'Local administrator', 'Domain user']),
        username          = dict(type='str'),
        password          = dict(type='str', no_log=True),
        domain            = dict(type='str'),
        tags              = dict(type='list'),
        additional_info   = dict(type='str'),
        parent_credential = dict(type='int'),
        validate_certs    = dict(type='bool', default=True)
    )


    required_if = [
        ["state", "present", ["device_address", "device_hostname", "type", "username", "identifier"]],
        ["state", "absent", ["identifier"]]
    ]

    module = AnsibleModule(
        argument_spec       = module_args,
        required_if         = required_if,
        supports_check_mode = False
    )

    authentication_token = authenticate(module)

    if (module.params['state'] == "present"):
        (changed, result, status_code) = ensure_credential(module, authentication_token)

    elif (module.params['state'] == "absent"):
        (changed, result, status_code) = ensure_absent_credential(module, authentication_token)

    module.exit_json(
        changed=changed,
        senhasegura_credential=result,
        status_code=status_code)


if __name__ == '__main__':
    main()
